File: .experiments/full187_f3_pure_face_k17_k18_vector_dual_6900.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <string>
#include <sys/resource.h>
#include <vector>

#include <givaro/modular.h>
#include <linbox/algorithms/polynomial-matrix/order-basis.h>
#include <linbox/matrix/polynomial-matrix.h>

namespace {
int64_t mul_mod(uint32_t a, uint32_t b, uint32_t p) {
    return static_cast<int64_t>((__int128)a * b % p);
}
int64_t neg_mod(uint32_t a, uint32_t p) { return a ? p - a : 0; }
}

int main(int argc, char** argv) {
    const std::string input = argc > 1 ? argv[1] :
        "/tmp/full187_f3_pure_face_k17_k18_vector_dual_6900.bin";
    std::ifstream in(input, std::ios::binary);
    if (!in) { std::cerr << "cannot open " << input << "\n"; return 2; }
    std::array<char,8> magic{};
    std::array<uint32_t,12> h{};
    in.read(magic.data(), magic.size());
    in.read(reinterpret_cast<char*>(h.data()), sizeof(h));
    if (std::string(magic.data(), 7) != "F3K4X2D" ||
        h[9] != 6 || h[10] != 4) {
        std::cerr << "bad header\n"; return 2;
    }
    const uint32_t prime=h[0], edeg=h[1], mdeg=h[2], rwidth=h[3];
    const uint32_t zwidth=h[4], outrows=h[5], lowbound=h[6];
    const uint32_t order=h[7], threshold=h[8], column_shift=h[11];
    std::array<uint32_t,4> bounds{}, initial_shift{};
    in.read(reinterpret_cast<char*>(bounds.data()), sizeof(bounds));
    in.read(reinterpret_cast<char*>(initial_shift.data()),
            sizeof(initial_shift));
    if (!in || mdeg != 18*edeg || lowbound+outrows != mdeg ||
        order != edeg+outrows || column_shift != 57193) {
        std::cerr << "inconsistent dimensions\n"; return 2;
    }

    // int64 storage is required for this 31-bit field.
    using Field = Givaro::Modular<int64_t>;
    using PMat = LinBox::PolynomialMatrix<Field, LinBox::PMType::polfirst>;
    Field F(prime);
    PMat series(F,4,2,order);
    std::vector<uint32_t> data(order);
    for (size_t which=0; which<6; ++which) {
        in.read(reinterpret_cast<char*>(data.data()),
                data.size()*sizeof(uint32_t));
        if (!in) { std::cerr << "truncated input\n"; return 2; }
        for (size_t k=0; k<order; ++k) {
            const uint32_t x=data[k];
            switch(which) {
            case 0: series.ref(0,0,k)=x; break;
            case 1: series.ref(1,0,k)=x; break;
            case 2: series.ref(0,1,k)=x; break;
            case 3: series.ref(1,1,k)=x; break;
            case 4: series.ref(2,1,k)=x; break;
            case 5: series.ref(3,1,k)=x; break;
            }
        }
    }

    std::vector<size_t> shift(initial_shift.begin(), initial_shift.end());
    PMat basis(F,4,4,order+1);
    LinBox::OrderBasis<Field> ob(F);
    const size_t det_degree=ob.PM_Basis(basis,series,order,shift);

    std::cout << "p_E_M_R_Z_Q_B_order " << prime << " " << edeg << " "
              << mdeg << " " << rwidth << " " << zwidth << " "
              << outrows << " " << lowbound << " " << order << "\n";
    std::cout << "threshold " << threshold << "\n";
    std::cout << "det_degree " << det_degree << "\n";
    std::cout << "initial_shift";
    for (auto x:initial_shift) std::cout << " " << x;
    std::cout << "\nfinal_shift";
    for (auto x:shift) std::cout << " " << x;
    std::cout << "\n";

    // The first column is x^column_shift times a primitive column and the
    // second is primitive independently (the C,S minor is a unit after
    // removing that power).  The determinant degree is therefore
    // (order-column_shift)+order.
    std::array<std::array<int64_t,4>,4> lead{};
    for(size_t i=0;i<4;i++) for(size_t j=0;j<4;j++) {
        if(shift[i] >= initial_shift[j]) {
            const size_t k=shift[i]-initial_shift[j];
            if(k < basis.size()) lead[i][j]=basis.get(i,j,k);
        }
    }
    int64_t lead_det=1;
    for(size_t col=0;col<4;col++) {
        size_t pivot=col;
        while(pivot<4 && lead[pivot][col]==0) ++pivot;
        if(pivot==4) { lead_det=0; break; }
        if(pivot!=col) {
            std::swap(lead[pivot],lead[col]);
            lead_det=lead_det ? prime-lead_det : 0;
        }
        const int64_t pv=lead[col][col];
        lead_det=static_cast<int64_t>((__int128)lead_det*pv%prime);
        int64_t inv=1, base=pv, exp=prime-2;
        while(exp) {
            if(exp&1) inv=static_cast<int64_t>((__int128)inv*base%prime);
            base=static_cast<int64_t>((__int128)base*base%prime);
            exp>>=1;
        }
        for(size_t row=col+1;row<4;row++) if(lead[row][col]) {
            const int64_t factor=static_cast<int64_t>(
                (__int128)lead[row][col]*inv%prime);
            for(size_t k=col;k<4;k++) {
                lead[row][k]=(lead[row][k]-static_cast<int64_t>(
                    (__int128)factor*lead[col][k]%prime))%prime;
                if(lead[row][k]<0) lead[row][k]+=prime;
            }
        }
    }
    size_t initial_sum=0, final_sum=0;
    for(size_t i=0;i<4;i++) {
        initial_sum += initial_shift[i];
        final_sum += shift[i];
    }
    const size_t shift_gain=final_sum-initial_sum;
    std::cout << "shift_gain_expected " << shift_gain << " "
              << 2*static_cast<size_t>(order)-column_shift << "\n";
    std::cout << "shift_leading_determinant " << lead_det << "\n";
    if(shift_gain != 2*static_cast<size_t>(order)-column_shift || lead_det==0) {
        std::cerr << "invalid shifted-basis certificate\n"; return 3;
    }

    const std::string basis_path=
        "/tmp/full187_f3_pure_face_k17_k18_vector_dual_basis_6900.bin";
    std::ofstream bout(basis_path,std::ios::binary);
    const std::array<char,8> bmagic={'F','3','K','4','X','2','B','\0'};
    bout.write(bmagic.data(),bmagic.size());
    const std::array<uint32_t,3> bh={prime,order,
                                     static_cast<uint32_t>(basis.size())};
    bout.write(reinterpret_cast<const char*>(bh.data()),sizeof(bh));
    for(auto x:initial_shift) { uint32_t y=x;
        bout.write(reinterpret_cast<const char*>(&y),sizeof(y)); }
    for(auto x:shift) { uint32_t y=static_cast<uint32_t>(x);
        bout.write(reinterpret_cast<const char*>(&y),sizeof(y)); }
    for(size_t i=0;i<4;i++) for(size_t j=0;j<4;j++)
        for(size_t k=0;k<basis.size();k++) {
            uint32_t x=static_cast<uint32_t>(basis.get(i,j,k));
            bout.write(reinterpret_cast<const char*>(&x),sizeof(x));
        }
    bout.close();
    std::cout << "basis_size " << basis.size() << "\n";
    std::cout << "basis_certificate " << basis_path << "\n";

    bool found=false;
    for(size_t i=0;i<4;i++) {
        std::array<long,4> degrees={-1,-1,-1,-1};
        for(size_t j=0;j<4;j++) for(size_t k=0;k<basis.size();k++)
            if(basis.get(i,j,k)) degrees[j]=static_cast<long>(k);
        std::cout << "row " << i << " shifted_degree " << shift[i]
                  << " component_degrees";
        for(auto x:degrees) std::cout << " " << x;
        if(shift[i] <= threshold) { found=true; std::cout << " CANDIDATE"; }
        std::cout << "\n";
    }
    rusage usage{};
    getrusage(RUSAGE_SELF,&usage);
    std::cout << "peak_rss_kib " << usage.ru_maxrss << "\n";
    if(found) {
        std::cout << "decision RED_NONZERO_COUPLED_VECTOR_DUAL\n";
    } else {
        std::cout << "decision GREEN_COUPLED_VECTOR_DUAL_INJECTIVE\n";
    }
    return 0;
}

File: .experiments/full187_f3_pure_face_k17_k18_vector_dual_6900.py
#!/usr/bin/env python3
"""Exact structured k=17/k=18 vector-dual gate for the Full187 F3 face.

This companion experiment starts from the 41,999-variable reduction in
``full187_f3_pure_face_graph_transform_6900.py``.  It constructs the literal
target modulus and the two modular multiplication operators.  The first mode
only probes the displacement structure of the remaining map; later modes emit
the compact simultaneous-approximant instance and verify its certificate.
"""

from __future__ import annotations

import argparse
from array import array
import hashlib
import json
from pathlib import Path
import resource
import struct
import sys

from flint import nmod_poly

sys.path.insert(0, str(Path(__file__).resolve().parent))
import full187_target_charge13_transposed_four_residue_gate_6900 as T


P = T.P
E_DEG = T.E
M_DEG = 18 * E_DEG
R_WIDTH = 40_240
Z_WIDTH = 1_759
OUT_ROWS = 98_684
LOW_BOUND = M_DEG - OUT_ROWS
A1 = 931_238_467
A2 = 95_217_577
DATA = Path("/tmp/full187_f3_pure_face_k17_k18_vector_dual_6900.bin")
BASIS = Path(
    "/tmp/full187_f3_pure_face_k17_k18_vector_dual_basis_6900.bin")


def sha_u32(values):
    h = hashlib.sha256()
    for start in range(0, len(values), 16384):
        chunk = values[start:start + 16384]
        h.update(struct.pack(f"<{len(chunk)}I", *chunk))
    return h.hexdigest()


def target_operators():
    instance = T.build_target_instance()
    h = nmod_poly(instance.xi_h_coefficients, P)
    rloc = nmod_poly(instance.xi_r_coefficients, P)
    e = nmod_poly(instance.xi_e_coefficients, P)
    g = h * rloc
    modulus = e ** 18
    u = g.pow_mod(58, modulus)
    g31 = g.pow_mod(31, modulus)
    gcd, v, _ = g31.xgcd(modulus)
    assert gcd == nmod_poly([1], P)
    v %= modulus
    assert (g31 * v) % modulus == nmod_poly([1], P)
    assert modulus.degree() == M_DEG
    return e, g, modulus, u, v


def coefficients(poly, length):
    return [int(poly[i]) for i in range(length)]


def exact_kernel_series():
    """Return the six nonzero entries of the reciprocal 4x2 system."""
    p18 = lambda n: __import__("math").prod(
        (n - root) % P for root in range(30, 72)) % P
    c18q, c18a = p18(2), p18(72)
    c17q, c17a = (2 - 29) * c18q % P, (72 - 29) * c18a % P
    assert (c17q, c17a, c18q, c18a, p18(29)) == (
        842_546_268, 1_616_860_076, 1_389_265_538,
        1_276_384_207, 1_276_384_207)
    assert c17q * pow(c17a, -1, P) % P == A1
    assert c18q * pow(c18a, -1, P) % P == A2
    assert (A1 - A2) % P == 836_020_890
    instance = T.build_target_instance()
    h = nmod_poly(instance.xi_h_coefficients, P)
    rloc = nmod_poly(instance.xi_r_coefficients, P)
    e = nmod_poly(instance.xi_e_coefficients, P)
    g = h * rloc
    modulus = e ** 18
    g31e = g.pow_mod(31, e)
    u58 = g.pow_mod(58, modulus)
    eg27 = (e * g.pow_mod(27, modulus)) % modulus
    eg58 = (e * u58) % modulus

    # Reverse at the exact ambient product degrees.  The first equality only
    # asks that its remainder have degree<R_WIDTH, giving order
    # 2*E_DEG-R_WIDTH.  The second asks that its E^18 remainder have degree
    # <LOW_BOUND, giving order E_DEG+OUT_ROWS.  Shift the first column by the
    # order difference to use one ordinary PM-basis call.
    order1 = 2 * E_DEG - R_WIDTH
    order2 = E_DEG + OUT_ROWS
    column_shift = order2 - order1
    assert (order1, order2, column_shift) == (123_222, 180_415, 57_193)
    g31star = g31e.reverse(degree=E_DEG - 1)
    estar = e.reverse(degree=E_DEG)
    u58star = u58.reverse(degree=M_DEG - 1)
    eg27star = eg27.reverse(degree=M_DEG - 1)
    eg58star = eg58.reverse(degree=M_DEG - 1)
    mstar = modulus.reverse(degree=M_DEG)

    # Exact orientation audit on sparse non-palindromic test polynomials.
    ty = nmod_poly([3, 5], P)
    tc = nmod_poly([7, 11], P)
    tz = nmod_poly([13, 17], P)
    ts = nmod_poly([19, 23], P)
    ty_star = ty.reverse(degree=E_DEG - 1)
    tc_star = tc.reverse(degree=E_DEG - 1)
    tz_star = tz.reverse(degree=Z_WIDTH - 1)
    ts_star = ts.reverse(degree=E_DEG - 1)
    direct1 = g31e * ty - e * tc
    reverse1 = (g31star.left_shift(1) * ty_star - estar * tc_star)
    assert reverse1 == direct1.reverse(degree=2 * E_DEG - 1)
    direct2 = (A2 * u58 * ty + (A1 - A2) * eg27 * tc
               + A1 * eg58 * tz - modulus * ts)
    reverse2 = (
        A2 * u58star.left_shift(1) * ty_star
        + (A1 - A2) * eg27star.left_shift(1) * tc_star
        + A1 * eg58star.left_shift(E_DEG - Z_WIDTH + 1) * tz_star
        - mstar * ts_star)
    assert reverse2 == direct2.reverse(degree=M_DEG + E_DEG - 1)
    arrays = (
        g31star.left_shift(column_shift + 1).truncate(order2),
        -estar.left_shift(column_shift).truncate(order2),
        (A2 * u58star.left_shift(1)).truncate(order2),
        ((A1 - A2) * eg27star.left_shift(1)).truncate(order2),
        (A1 * eg58star.left_shift(E_DEG - Z_WIDTH + 1)).truncate(order2),
        -mstar.truncate(order2),
    )
    # Variables are fixed-length reversals (Y,C,Z,S).
    bounds = (E_DEG, E_DEG, Z_WIDTH, E_DEG)
    threshold = E_DEG - 1
    shifts = tuple(threshold - (bound - 1) for bound in bounds)
    assert shifts == (0, 0, 79_972, 0)
    assert all(poly.degree() < order2 for poly in arrays)
    return arrays, bounds, shifts, order2, threshold, order1, column_shift


def generate():
    arrays, bounds, shifts, order, threshold, order1, column_shift = (
        exact_kernel_series())
    with DATA.open("wb") as stream:
        stream.write(struct.pack(
            "<8s12I", b"F3K4X2D", P, E_DEG, M_DEG, R_WIDTH,
            Z_WIDTH, OUT_ROWS, LOW_BOUND, order, threshold,
            len(arrays), len(bounds), column_shift))
        stream.write(struct.pack(f"<{len(bounds)}I", *bounds))
        stream.write(struct.pack(f"<{len(shifts)}I", *shifts))
        for poly in arrays:
            values = coefficients(poly, order)
            stream.write(struct.pack(f"<{len(values)}I", *values))
    print(json.dumps({
        "output": str(DATA),
        "output_bytes": DATA.stat().st_size,
        "p_Edeg_Mdeg_rWidth_zWidth_outRows_lowBound": (
            P, E_DEG, M_DEG, R_WIDTH, Z_WIDTH, OUT_ROWS, LOW_BOUND),
        "bounds_reversed_Y_C_Z_S": bounds,
        "initial_shifts": shifts,
        "effective_orders_columnShift_threshold": (
            order1, order, column_shift, threshold),
        "array_degrees": tuple(poly.degree() for poly in arrays),
        "array_sha256_u32": tuple(
            sha_u32(coefficients(poly, order)) for poly in arrays),
        "binary_sha256": hashlib.sha256(DATA.read_bytes()).hexdigest(),
        "peak_rss_kib": resource.getrusage(resource.RUSAGE_SELF).ru_maxrss,
        "decision": "GENERATED_EXACT_RECIPROCAL_4_BY_2_PM_INSTANCE",
    }, indent=2, sort_keys=True))


def read_u32_poly(stream, length):
    raw = array("I")
    raw.fromfile(stream, length)
    if sys.byteorder != "little":
        raw.byteswap()
    assert len(raw) == length
    return nmod_poly(list(raw), P)


def verify_basis():
    """Independently verify every PM-basis product and degree certificate."""
    with DATA.open("rb") as stream:
        header = struct.unpack("<8s12I", stream.read(56))
        magic, p, edeg, mdeg, rwidth, zwidth, outrows, lowbound, order, \
            threshold, count, bound_count, column_shift = header
        assert magic == b"F3K4X2D\0"
        assert (p, edeg, mdeg, rwidth, zwidth, outrows, lowbound) == (
            P, E_DEG, M_DEG, R_WIDTH, Z_WIDTH, OUT_ROWS, LOW_BOUND)
        assert (count, bound_count, column_shift) == (6, 4, 57_193)
        bounds = struct.unpack("<4I", stream.read(16))
        initial_shift = struct.unpack("<4I", stream.read(16))
        arrays = [read_u32_poly(stream, order) for _ in range(6)]
        assert not stream.read(1)
    zero = nmod_poly([], P)
    series = (
        (arrays[0], arrays[2]),
        (arrays[1], arrays[3]),
        (zero, arrays[4]),
        (zero, arrays[5]),
    )

    with BASIS.open("rb") as stream:
        magic2, p2, order2, size = struct.unpack("<8s3I", stream.read(20))
        assert magic2 == b"F3K4X2B\0"
        assert (p2, order2) == (P, order)
        initial2 = struct.unpack("<4I", stream.read(16))
        final_shift = struct.unpack("<4I", stream.read(16))
        assert initial2 == initial_shift
        actual_final = []
        lead = [[0] * 4 for _ in range(4)]
        zero_checks = 0
        component_degrees = []
        for i in range(4):
            row = [read_u32_poly(stream, size) for _ in range(4)]
            degrees = tuple(poly.degree() for poly in row)
            component_degrees.append(degrees)
            shifted_degree = max(
                (degree + initial_shift[j]
                 for j, degree in enumerate(degrees) if degree >= 0),
                default=-1)
            actual_final.append(shifted_degree)
            for j, poly in enumerate(row):
                k = final_shift[i] - initial_shift[j]
                lead[i][j] = int(poly[k]) if 0 <= k < size else 0
            for col in range(2):
                product = sum(
                    (row[j] * series[j][col] for j in range(4)), zero)
                assert not product.truncate(order)
                zero_checks += 1
            del row
        assert not stream.read(1)

    assert tuple(actual_final) == final_shift
    effective_degree = 2 * order - column_shift
    assert sum(final_shift) - sum(initial_shift) == effective_degree
    det = 1
    for col in range(4):
        pivot = next(i for i in range(col, 4) if lead[i][col])
        if pivot != col:
            lead[col], lead[pivot] = lead[pivot], lead[col]
            det = -det % P
        pv = lead[col][col]
        det = det * pv % P
        inv = pow(pv, -1, P)
        for row in range(col + 1, 4):
            factor = lead[row][col] * inv % P
            lead[row] = [
                (x - factor * y) % P
                for x, y in zip(lead[row], lead[col])]
    assert det
    found = min(final_shift) <= threshold
    print(json.dumps({
        "basis_file": str(BASIS),
        "basis_file_bytes": BASIS.stat().st_size,
        "basis_sha256": hashlib.sha256(BASIS.read_bytes()).hexdigest(),
        "bounds_reversed_Y_C_Z_S": bounds,
        "initial_shift": initial_shift,
        "final_shift": final_shift,
        "component_degrees": component_degrees,
        "threshold": threshold,
        "minimum_shifted_degree_gap": min(final_shift) - threshold,
        "column_orders_and_shift_gain": (
            order-column_shift, order,
            sum(final_shift)-sum(initial_shift)),
        "shift_leading_determinant": det,
        "approximant_zero_checks": zero_checks,
        "peak_rss_kib": resource.getrusage(resource.RUSAGE_SELF).ru_maxrss,
        "decision": (
            "RED_NONZERO_COUPLED_VECTOR_DUAL" if found else
            "CERTIFIED_GREEN_COUPLED_VECTOR_DUAL_INJECTIVE"),
    }, indent=2, sort_keys=True))


def tail(poly):
    return tuple(int(poly[i]) for i in range(LOW_BOUND, M_DEG))


def r_column(j, e, modulus, u, v):
    y = v.left_shift(j) % modulus
    y0 = y % e
    gq = A1 * y0 + (A2 - A1) * y
    return tail((u * gq) % modulus)


def z_column(j, e, modulus, u):
    gq = A1 * e.left_shift(j)
    return tail((u * gq) % modulus)


def shifted_equal(a, b, direction):
    if direction == "down":
        return a[1:] == b[:-1]
    if direction == "up":
        return a[:-1] == b[1:]
    raise AssertionError(direction)


def probe():
    e, g, modulus, u, v = target_operators()
    rcols = [r_column(j, e, modulus, u, v) for j in range(3)]
    zcols = [z_column(j, e, modulus, u) for j in range(3)]
    receipt = {
        "p_Edeg_Mdeg_rWidth_zWidth_outRows_lowBound": (
            P, E_DEG, M_DEG, R_WIDTH, Z_WIDTH, OUT_ROWS, LOW_BOUND),
        "degrees_E_G_M_U_Ginv31": (
            e.degree(), g.degree(), modulus.degree(), u.degree(), v.degree()),
        "r_column_sha256_u32": tuple(sha_u32(x) for x in rcols),
        "z_column_sha256_u32": tuple(sha_u32(x) for x in zcols),
        "adjacent_shift_tests": {
            "r_down": tuple(shifted_equal(rcols[j], rcols[j + 1], "down")
                            for j in range(2)),
            "r_up": tuple(shifted_equal(rcols[j], rcols[j + 1], "up")
                          for j in range(2)),
            "z_down": tuple(shifted_equal(zcols[j], zcols[j + 1], "down")
                            for j in range(2)),
            "z_up": tuple(shifted_equal(zcols[j], zcols[j + 1], "up")
                          for j in range(2)),
        },
        "peak_rss_kib": resource.getrusage(resource.RUSAGE_SELF).ru_maxrss,
        "decision": "STRUCTURE_PROBE_ONLY",
    }
    print(json.dumps(receipt, indent=2, sort_keys=True))


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--probe", action="store_true")
    parser.add_argument("--generate", action="store_true")
    parser.add_argument("--verify-basis", action="store_true")
    args = parser.parse_args()
    if args.probe:
        probe()
    elif args.generate:
        generate()
    elif args.verify_basis:
        verify_basis()
    else:
        parser.error("select --probe or --generate")


if __name__ == "__main__":
    main()
